### ISSUE TITLE
Mapped array type with generic member as rest parameters prevents spread of same type

### ISSUE THREAD

**Issue by aleclarson** [Microsoft/TypeScript#29763]
**TypeScript Version:**  `3.4.0-dev.20190202` and `3.3.1`

**Search Terms:** "generic rest", "generic spread"

**Code**

```ts
type Foo<T extends any[]> = {
    [P in keyof T]: T[P]
}

function test<T extends any[]>(fn: (...args: Foo<T>) => void) {
    const arr: Foo<T> = {} as any
    fn(...arr) // Error: Argument of type 'any[]' is not assignable to parameter of type 'Foo<T>'
}
```

**Expected behavior:** No error.

**Actual behavior:** Argument of type 'any[]' is not assignable to parameter of type 'Foo<T>'

**Playground Link:** [Click here][1]

**Related Issues:** None

[1]: http://www.typescriptlang.org/play/#src=type%20Foo%3CT%20extends%20any%5B%5D%3E%20%3D%20%7B%0A%20%20%20%20%5BP%20in%20keyof%20T%5D%3A%20T%5BP%5D%0A%7D%0A%0Afunction%20test%3CT%20extends%20any%5B%5D%3E(fn%3A%20(...args%3A%20Foo%3CT%3E)%20%3D%3E%20void)%20%7B%0A%20%20%20%20const%20arr%3A%20Foo%3CT%3E%20%3D%20%7B%7D%20as%20any%0A%20%20%20%20fn(...arr)%0A%7D%0A

**Comment by RyanCavanaugh** [comment]
I'm not sure we're going to be able to traffic the mapped type through the arguments rest to figure out that this is actually valid.

What's the actual use case for why `test` is written this way?

**Comment by aleclarson** [comment]
The mapped type is "unboxing" values from a generic type, and the given function is an event listener.

**Comment by weswigham** [comment]
So, two issues here. One: A spread (`...` in an argument position) always unwraps its argument and repackages it into a real array type - this is to somewhat match what happens at runtime (where whatever you spread in is unpacked into an arguments list). Because of this, an expression like `...(arr as Foo<T>)` will _never_ be assignable to a `Foo<T>` since it should be transformed into to a `Foo<T>[number][]` instead (which isn't actually compatible with a `Foo<T>`). We probably shouldn't let you define rest types like this, if I'm honest. Two: we don't actually _do_ that with generics - we resolve to the constraint instead, which is why the error is `any[] not assignable` and not the more useful `Foo<T>[number][] not assignable to Foo<T>`. I can fix the second and will open a PR; the first, however is an issue with the assumptions your own code is making. A spread does not actually capture the variant of array type input to it.
